Compiler/src/Compiler.jl

```diff
@@ -167,7 +180,7 @@ include("optimize.jl")
 include("bootstrap.jl")
 include("reflection_interface.jl")
 
-if isdefined(Base, :IRShow)
+if isdefined(Base, :end_base_include)
```

Comment:
@Keno In some cases, a precompilation cache for Compiler.jl is created without `Compiler.IRShow` being used by Cthulhu.jl when precompiling it, causing the precompilation error. I suspect it might be related to the check on line 21. Do you have any ideas? Should IRShow.jl be added to `Base._compiler_require_dependencies`?

Reply:
That's a fair point. Perhaps
```
keno@demeter1:~/julia-clean$ git diff
diff --git a/base/show.jl b/base/show.jl
index e332cf521a..e6f85414a1 100644
--- a/base/show.jl
+++ b/base/show.jl
@@ -2823,7 +2823,7 @@ end
 
 module IRShow
     using ..Compiler: Compiler
-    Base.include(IRShow, Base.strcat(Base.BUILDROOT, "../usr/share/julia/Compiler/src/ssair/show.jl"))
+    Compiler.include(IRShow, "ssair/show.jl")
 
     const __debuginfo = Dict{Symbol, Any}(
         # :full => src -> statementidx_lineinfo_printer(src), # and add variable slot information
```
would do the trick?

It's a little bit strange, because in the base version, it's not actually in the compiler module (although maybe it should be? - there isn't really any reason not to), but I think it should work.

Reply:
That diff alone doesn’t seem to be sufficient. It looks like the issue can be resolved by modifying `Compiler` to define `IRShow` within it
```diff
diff --git a/Compiler/src/Compiler.jl b/Compiler/src/Compiler.jl
index cdc3be1b4f..376721da46 100644
--- a/Compiler/src/Compiler.jl
+++ b/Compiler/src/Compiler.jl
@@ -180,12 +180,12 @@ include("optimize.jl")
 include("bootstrap.jl")
 include("reflection_interface.jl")
 
-if isdefined(Base, :end_base_include)
-    @eval module IRShow
-        using ..Compiler: Compiler
-        # During bootstrap, Base will later include this into its own "IRShow module"
-        Compiler.include(IRShow, "ssair/show.jl")
-    end
+module IRShow end
+if !isdefined(Base, :end_base_include)
+    # During bootstrap, skip including this file and defer it to base/show.jl to include later
+else
+    # When this module is loaded as the standard library, include this file as usual
+    include(IRShow, "ssair/show.jl")
 end
 
 end # baremodule Compiler
diff --git a/Compiler/src/ssair/show.jl b/Compiler/src/ssair/show.jl
index 22394c84eb..b9ed220d59 100644
--- a/Compiler/src/ssair/show.jl
+++ b/Compiler/src/ssair/show.jl
@@ -1,7 +1,8 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-# This file is not loaded into `Core.Compiler` but rather loaded into the context of
-# `Base.IRShow` and thus does not participate in bootstrapping.
+# This file does not participate in bootstrapping, but is included in the system image by
+# being loaded from `base/show.jl`. Compiler.jl as the standard library will simply include
+# this file in the context of `Compiler.IRShow`.
 
 using Base, Core.IR
 
diff --git a/base/show.jl b/base/show.jl
index 6fa3f0cf8c..6856d37e36 100644
--- a/base/show.jl
+++ b/base/show.jl
@@ -2821,10 +2821,8 @@ function show(io::IO, vm::Core.TypeofVararg)
     end
 end
 
-module IRShow
-    using ..Compiler: Compiler
-    Base.include(IRShow, Base.strcat(Base.BUILDROOT, "../usr/share/julia/Compiler/src/ssair/show.jl"))
-end
+@eval Compiler include(IRShow, "ssair/show.jl") # define `show` for the compiler types
+const IRShow = Compiler.IRShow # an alias for compatibility
 
 function show(io::IO, src::CodeInfo; debuginfo::Symbol=:source)
     # Fix slot names and types in function body
```

Reply:
Yeah, that's what I meant with my second comment. I don't really see any reason not to do that.